File: 234. Palindrome Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if (head == NULL) return true;
        ListNode *fast, *slow;
        fast = head; slow = head;
        while (fast->next && fast->next->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        slow->next = reverse(slow->next);
        slow = slow->next;
        
        while (slow) {
            if (head->val != slow->val)
                return false;
            head = head->next;
            slow = slow->next;
        }
        
        return true;
    }
    
    ListNode* reverse(ListNode* head) {
        ListNode *next = NULL, *pre = NULL ;
        while (head) {
            next = head->next;
            head->next = pre;
            pre = head;
            head = next;
        }
        return pre;
    }
};

File: python/234. Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head:
            return True
        cnt = 0
        dummy = ListNode(0)
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            cnt += 1
        
        p = slow
        tail = p
        while p:
            tmp = p.next
            p.next = dummy.next
            dummy.next = p
            p = tmp
            cnt += 1
        tail.next = None
        # print cnt
        # if cnt&1:
        p = dummy.next
        while p and head:
            if p.val != head.val:
                return False
            p = p.next
            head = head.next
        return True
